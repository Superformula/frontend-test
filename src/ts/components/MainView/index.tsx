// Imports
// -------

// Libraries
import React, {useState, useCallback} from 'react';
// Components
import CardGallery from '../CardGallery';
import CategoriesDropDown from "../CategoriesDropDown";
import Hero from '../Hero';
import FilterBy from '../FilterBy';
import Checkbox from "../Checkbox";
import DropDownMenu from '../DropDownMenu';
import Button from "../Button";
// GraphQL
import getRestaurantsQuery from '../../graphql/yelp/queries/getRestaurants';
// Hooks
import requestData from "../../hooks/requestData";
// Styles
import style from './style.module.css';

// Internal
// --------

export interface MainViewProps {
}

const MainView: React.FC<MainViewProps> = ({}) => {
    //
    // State

    const [openNow, setOpenNow] = useState(false);
    const [priceMenuSelected, setPriceMenuSelected] = useState(false);
    const [priceSelected, setPriceSelected] = useState<null | string>(null);
    const [categories, setCategories] = useState([]);
    const [categorySelected, setCategorySelected] = useState<null | string>(null);

    //
    // Callback

    const openNowCallback = useCallback(() => {
        setOpenNow(!openNow);
    }, [openNow, setOpenNow]);

    const priceMenuSelectedCallback = useCallback(() => {
        setPriceMenuSelected(!priceMenuSelected);
    }, [priceMenuSelected, setPriceMenuSelected]);

    const priceSelectedCallback = useCallback((price: string) => {
        setPriceSelected(price);
    }, [setPriceSelected]);

    const categorySelectedCallback = useCallback((category: string) => {
        setCategorySelected(category);
    }, [setCategorySelected]);

    const clearAllCallback = useCallback(() => {
        setOpenNow(false);
        setPriceSelected(null);
        setCategorySelected(null);
    }, [setOpenNow, setPriceSelected, setCategorySelected]);

    //
    // Render

    const mock = [{id: 'id', imgSrc: 'http://via.placeholder.com/500', title: 'Title', rating: 2, category: 'seafood', price: 2, status: 'open' },
        {id: 'id', imgSrc: 'http://via.placeholder.com/500', title: 'Title', rating: 2.5, category: 'seafood', price: 2, status: 'open' },
        {id: 'id', imgSrc: 'http://via.placeholder.com/500', title: 'Title', rating: 3, category: 'seafood', price: 2, status: 'open' },
        {id: 'id', imgSrc: 'http://via.placeholder.com/500', title: 'Title', rating: 2, category: 'seafood', price: 2, status: 'open' },
        {id: 'id', imgSrc: 'http://via.placeholder.com/500', title: 'Title', rating: 2, category: 'seafood', price: 2, status: 'open' },
        {id: 'id', imgSrc: 'http://via.placeholder.com/500', title: 'Title', rating: 2, category: 'seafood', price: 2, status: 'open' }];

    const disableClearAll = !openNow && priceSelected === null && categorySelected === null;

    return (
        <div>
            <Hero title="Restaurants"/>
            <div className={ style.filterSection }>
                <FilterBy clearAll={ clearAllCallback } disableClearAll={ disableClearAll }>
                    <Checkbox role="radiobutton" checked={ openNow } onClick={ openNowCallback }>Open Now</Checkbox>
                    <DropDownMenu text="Price" selected={ priceMenuSelected } onClick={ priceMenuSelectedCallback } dimensions={{ width: 97, height: 32 }} option={ priceSelected } setOption={ priceSelectedCallback }>
                        { ['All', '$', '$$', '$$$', '$$$$'] }
                    </DropDownMenu>
                    <CategoriesDropDown categorySelected={ categorySelected } setCategories={ setCategories } categorySelectedCallback={ categorySelectedCallback }>
                        { categories }
                    </CategoriesDropDown>
                </FilterBy>
            </div>
            <h1>All Restaurants</h1>
            <CardGallery data={ mock } />
            <div className={ style.loadMoreButton }>
                <Button text='Load More' tier='secondary' size='large' />
            </div>
        </div>
    );
}

// Exports
// -------

export default MainView;
