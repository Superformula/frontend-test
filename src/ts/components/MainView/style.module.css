.filterSection {
    margin: 0 -64px;
    border-top: 1px solid #e6e6e6;
    border-bottom: 1px solid #e6e6e6;
    padding: 21px 64px;
}

.loadMoreButton {
    display: flex;
    justify-items: center;
    text-align: center;
}

.loadMoreButton button {
    margin: 0 auto;
}
