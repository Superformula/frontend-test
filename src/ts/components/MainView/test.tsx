// Imports
// -------

// Libraries
import React from 'react';
import renderer from 'react-test-renderer';
// Components
import MainView from './index';

// Mocks
// -----

jest.mock("../CardGallery", () => 'CardGallery');
jest.mock('../CategoriesDropDown', () => 'CategoriesDropDown');
jest.mock("../Hero", () => 'Hero');
jest.mock('../FilterBy', () => 'FilterBy');
jest.mock("../Checkbox", () => 'Checkbox');
jest.mock('../DropDownMenu', () => 'DropDownMenu');
jest.mock('../Button', () => 'Button');

// Internal
// --------

const getComponent = (props = {}) => renderer.create(
    <MainView {...props} />,
);

// Tests
// -----

describe('src/ts/components/MainView', () => {
    let component;

    describe('Rendering', () => {
        it('renders default correctly', () => {
            component = getComponent();
            expect(component).toMatchSnapshot();
        });
    });
});
