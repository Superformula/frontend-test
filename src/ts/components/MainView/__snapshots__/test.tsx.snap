// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/ts/components/MainView Rendering renders default correctly 1`] = `
<div>
  <Hero
    title="Restaurants"
  />
  <div
    className="filterSection"
  >
    <FilterBy
      clearAll={[Function]}
      disableClearAll={true}
    >
      <Checkbox
        checked={false}
        onClick={[Function]}
        role="radiobutton"
      >
        Open Now
      </Checkbox>
      <DropDownMenu
        dimensions={
          Object {
            "height": 32,
            "width": 97,
          }
        }
        onClick={[Function]}
        option={null}
        selected={false}
        setOption={[Function]}
        text="Price"
      >
        All
        $
        $$
        $$$
        $$$$
      </DropDownMenu>
      <CategoriesDropDown
        categorySelected={null}
        categorySelectedCallback={[Function]}
        setCategories={[Function]}
      />
    </FilterBy>
  </div>
  <h1>
    All Restaurants
  </h1>
  <CardGallery
    data={
      Array [
        Object {
          "category": "seafood",
          "id": "id",
          "imgSrc": "http://via.placeholder.com/500",
          "price": 2,
          "rating": 2,
          "status": "open",
          "title": "Title",
        },
        Object {
          "category": "seafood",
          "id": "id",
          "imgSrc": "http://via.placeholder.com/500",
          "price": 2,
          "rating": 2.5,
          "status": "open",
          "title": "Title",
        },
        Object {
          "category": "seafood",
          "id": "id",
          "imgSrc": "http://via.placeholder.com/500",
          "price": 2,
          "rating": 3,
          "status": "open",
          "title": "Title",
        },
        Object {
          "category": "seafood",
          "id": "id",
          "imgSrc": "http://via.placeholder.com/500",
          "price": 2,
          "rating": 2,
          "status": "open",
          "title": "Title",
        },
        Object {
          "category": "seafood",
          "id": "id",
          "imgSrc": "http://via.placeholder.com/500",
          "price": 2,
          "rating": 2,
          "status": "open",
          "title": "Title",
        },
        Object {
          "category": "seafood",
          "id": "id",
          "imgSrc": "http://via.placeholder.com/500",
          "price": 2,
          "rating": 2,
          "status": "open",
          "title": "Title",
        },
      ]
    }
  />
  <div
    className="loadMoreButton"
  >
    <Button
      size="large"
      text="Load More"
      tier="secondary"
    />
  </div>
</div>
`;
